### ISSUE TITLE
database from stripped version of hpcstruct confuses hpcviewer

### ISSUE THREAD

**Issue by jmellorcrummey** [HPCToolkit/hpcviewer#71]
Try the bottom-up view and use flame to open the call paths from unknown procedure. It has unknown procedure as its parent. When you try to open them, things go wrong.

This manifests itself on both mac and x86_64 linux, which is where I've tried so far, with recent hpcviewers.

Is the problem with hpcviewer, or the fact that hpcprof generates <unknown procedure> for all procedures in the stripped binary without a virtual address tag?

**Comment by jmellorcrummey** [comment]
image from mac

![image](https://user-images.githubusercontent.com/3926112/74492224-d6418a80-4e93-11ea-9d3a-5a9406df7f71.png)

image from Linux

![image](https://user-images.githubusercontent.com/3926112/74492257-f2ddc280-4e93-11ea-9d54-32a0e9044427.png)

**Comment by jmellorcrummey** [comment]
one more bad image from linux (like the mac version when the third unknown procedure parent is opened)

![image](https://user-images.githubusercontent.com/3926112/74492303-17d23580-4e94-11ea-9a67-71f445c2fd54.png)

**Comment by jmellorcrummey** [comment]
database can be found here:  https://rice.box.com/s/dz9zfmdtgsgo68ibxiapyhdbwbzcep05

**Comment by laksono** [comment]
Looks like this is pure hpcviewer problem. 
There is an inconsistency between the data and the lazy tree content provider. Not sure why a widget tree item is suddenly disposed.

**Comment by laksono** [comment]
Error stack (widget disposed)
```
SWT.error(int, Throwable, String) line: 4535	
SWT.error(int, Throwable) line: 4448	
SWT.error(int) line: 4419	
TreeItem(Widget).error(int) line: 799	
TreeItem(Widget).checkWidget() line: 588	
TreeItem.setItemCount(int) line: 1428	
ScopeTreeViewer(TreeViewer).lambda$2(Object, int) line: 383	
269685385.run() line: not available [local variables unavailable]	
ScopeTreeViewer(StructuredViewer).preservingSelection(Runnable, boolean) line: 1458	
ScopeTreeViewer(TreeViewer).preservingSelection(Runnable, boolean) line: 354	
ScopeTreeViewer(StructuredViewer).preservingSelection(Runnable) line: 1419	
ScopeTreeViewer(TreeViewer).setChildCount(Object, int) line: 375	
CallerViewContentProvider(AbstractContentProvider).updateChildCount(Object, int) line: 71	
CallerViewContentProvider(AbstractContentProvider).updateElement(Object, int) line: 61	
ScopeTreeViewer(TreeViewer).virtualLazyUpdateWidget(Widget, int) line: 985	
ScopeTreeViewer(TreeViewer).lambda$1(Event) line: 255	
1035394386.handleEvent(Event) line: not available	
EventTable.sendEvent(Event) line: 84	
Display.sendEvent(EventTable, Event) line: 4256	
Tree(Widget).sendEvent(Event) line: 1501	
Tree(Widget).sendEvent(int, Event, boolean) line: 1524	
Tree(Widget).sendEvent(int, Event) line: 1509	
Tree.checkData(TreeItem) line: 329	
TreeItem.getText(int) line: 870	
TreeViewerRow.getText(int) line: 88	
ViewerCell.getText() line: 111	
StyledScopeLabelProvider(DelegatingStyledCellLabelProvider).update(ViewerCell) line: 114	
TreeViewerColumn(ViewerColumn).refresh(ViewerCell) line: 141	
ScopeTreeViewer(AbstractTreeViewer).doUpdateItem(Item, Object) line: 949	
AbstractTreeViewer$UpdateItemSafeRunnable.run() line: 118	
SafeRunner.run(ISafeRunnable) line: 42	
JFaceUtil$1.run(ISafeRunnable) line: 50	
SafeRunnable.run(ISafeRunnable) line: 173	
ScopeTreeViewer(AbstractTreeViewer).doUpdateItem(Widget, Object, boolean) line: 1029	
StructuredViewer$UpdateItemSafeRunnable.run() line: 474	
SafeRunner.run(ISafeRunnable) line: 42	
JFaceUtil$1.run(ISafeRunnable) line: 50	
SafeRunnable.run(ISafeRunnable) line: 173	
ScopeTreeViewer(StructuredViewer).updateItem(Widget, Object) line: 2170	
ScopeTreeViewer(TreeViewer).replace(Object, int, Object) line: 472	
CallerViewContentProvider(AbstractContentProvider).updateElement(Object, int) line: 60	
ScopeTreeViewer(TreeViewer).virtualLazyUpdateWidget(Widget, int) line: 985	
ScopeTreeViewer(TreeViewer).lambda$1(Event) line: 255	
1035394386.handleEvent(Event) line: not available	
EventTable.sendEvent(Event) line: 84	
Display.sendEvent(EventTable, Event) line: 4256	
Tree(Widget).sendEvent(Event) line: 1501	
Tree(Widget).sendEvent(int, Event, boolean) line: 1524	
Tree(Widget).sendEvent(int, Event) line: 1509	
Tree.checkData(TreeItem) line: 329	
Tree.outlineView_objectValueForTableColumn_byItem(long, long, long, long, long) line: 2147	
Display.windowProc(long, long, long, long, long) line: 6055	
OS.objc_msgSendSuper(objc_super, long, NSRect) line: not available [native method]	
Tree(Widget).drawRect(long, long, NSRect) line: 758	
Display.windowProc(long, long, long) line: 5711	
OS.objc_msgSend(long, long) line: not available [native method]	
SWTOutlineView(NSView).displayIfNeeded() line: 151	
Tree(Control).update(boolean) line: 5124	
Tree(Control).update() line: 5104	
ScopeTreeViewer(TreeViewer).setExpanded(Item, boolean) line: 292	
ScopeTreeViewer(AbstractTreeViewer).internalExpandToLevel(Widget, int) line: 1762	
ScopeTreeViewer(AbstractTreeViewer).expandToLevel(Object, int) line: 1071	
BaseScopeViewActions(ScopeViewActions).getHotCallPath(Scope, BaseMetric, int, TreePath, ScopeViewActions$HotCallPath) line: 140	
BaseScopeViewActions(ScopeViewActions).getHotCallPath(Scope, BaseMetric, int, TreePath, ScopeViewActions$HotCallPath) line: 163	
BaseScopeViewActions(ScopeViewActions).getHotCallPath(Scope, BaseMetric, int, TreePath, ScopeViewActions$HotCallPath) line: 163	
BaseScopeViewActions(ScopeViewActions).showHotCallPath() line: 296	
```

**Comment by laksono** [comment]
I think this is the tip of the iceberg. There is an issue with the way we construct "dynamic" bottom-up tree.

- When identical paths are merged, the cost associated with the "merged" node can be incorrect. In the snapshot below, the cost of the selected node (`<unknown procedure>`) is 43.5% while the cost of the parent is 25.2%. I believe the selected node consists of two merged nodes with identical paths: one with a cost of 25%, the other one is 43.5%.

![image](https://user-images.githubusercontent.com/8781032/74687487-1bb6ce00-519a-11ea-8aaa-e1bc213787ec.png)

- Two other related problems of cost attribution:
  - within Dyninst's `parse_vanilla` (cost: 34.1%) and its only child: `parse` (cost is suddenly 2.3%) functions (see the selected line).
  - within the merged recursive functions (`GOMP_task` and Dyninst's `launch_work`) in the bottom-up view. 

![image](https://user-images.githubusercontent.com/8781032/74690369-76edbe00-51a4-11ea-99af-ccb98c350e22.png)

**Comment by laksono** [comment]
Perhaps there is an issue with object references in bottom-up tree.
- Originally, expanding Dyninst's `parse_frames` procedure has no issue with the cost attribution: the cost is consistenly 34.1%:

![image](https://user-images.githubusercontent.com/8781032/74691863-37c26b80-51aa-11ea-992c-07d6f354d4ed.png)

- However, if we expand `parser_frames` until `Parser::parse` function that has 2.3% of total CPUTIME, suddenly the cost of Dyninst's `Parser::parse` (in another tree branch) also changes from 34.1% to 2.3%:

![image](https://user-images.githubusercontent.com/8781032/74692390-896bf580-51ac-11ea-88ae-f8dd915a7e7d.png)

It seems we use exactly the same object for the same node from different contexts.

**Comment by laksono** [comment]
This issue consists of two bugs:
- crashes or blank rows in bottom-up view (see the original problem)
- Incorrect cost attribution (see one of my comments above)

The first one is due to unreliable flat ID used in tree node's hashcode method, and fixed in commit 3636195a4e3dec10e7f859e62e26836624fb6fd2

The second problem occurs due to a combination of recursive functions `<unknown procedure>` and stripped functions.

![image](https://user-images.githubusercontent.com/8781032/74777356-ef18ba00-525e-11ea-95f7-1c40013cc88d.png)

The above snapshot shows that the selection node (`<unknown procedure>`) has an inclusive cost of 2.2%, but the exclusive cost is 20.7%. The exclusive cost is correct, but the inclusive cost is wrong.
This node is a merge of 64 `<unknown procedure>` and its inclusive cost should sum the two outermost `<unknown procedure>` nodes from the top-down tree, i.e. 97.8% + 2.2%. But it only computes the 2.2% node. 

![image](https://user-images.githubusercontent.com/8781032/74782351-e331f580-5268-11ea-828e-cf17f1fb32c8.png)
 
It seems the algorithm doesn't work for merging two recursive functions.

**Comment by laksono** [comment]
I think we can close this issue since we fixed struct-id (or flat-id) in hpcprof.
It's too tricky to handle old databases with redundant struct-id. In this case, the viewer will display incorrect costs in the bottom-up tree.
